pyccel/parser/semantic.py

```diff
@@ -1188,9 +1192,109 @@ def _handle_function(self, expr, func, args, is_method = False):
+        """
+        input_args = [a for a in args if a.keyword is None]
+        nargs = len(input_args)
+        for ka in func_args[nargs:]:
+            key = ka.name
+            relevant_args = [a for a in args[nargs:] if a.keyword == key]
+            n_relevant_args = len(relevant_args)
+            assert n_relevant_args <= 1
+            if n_relevant_args == 0 and ka.has_default:
+                input_args.append(ka.default_call_arg)
+            elif n_relevant_args == 1:
+                input_args.append(relevant_args[0])
+
+        return input_args
+
+    def _annotate_the_called_function_def(self, old_func, function_call=None):
+        """
+        Annotate the called FunctionDef.
+
+        Annotate the called FunctionDef.
+
+        Parameters
+        ----------
+        old_func : FunctionDef|Interface
+           The function that needs to be annotated.
+
+        function_call : list,optional
+           The list of the call arguments.
+
+        Returns
+        -------
+        func: FunctionDef|Interface
+            The new annotated function.
+        """
+        # The function call might be in a completely different scope from the FunctionDef
+        # Store the current scope and go to the parent scope of the FunctionDef
+        old_scope            = self._scope
+        old_current_function = self._current_function
+        names = []
+        sc = old_func.scope if isinstance(old_func, FunctionDef) else old_func.syntactic_node.scope
+        while sc.parent_scope is not None:
+            sc = sc.parent_scope
+            if not sc.name is None:
+                names.append(sc.name)
+        names.reverse()
+        if names:
+            self._current_function = DottedName(*names) if len(names)>1 else names[0]
+        else:
+            self._current_function = None
+
+        while names:
+            sc = sc.sons_scopes[names[0]]
+            names = names[1:]
+
+        # Set the Scope to the FunctionDef's parent Scope and annotate the old_func
+        self._scope = sc
+        self._visit_FunctionDef(old_func, annotate=True, function_call=function_call)
+        # Retreive the annotated function
+        func = self.scope.find(old_func.name, 'functions')
+        # Add the Module of the imported function to the new function
+        if old_func.is_imported:
+            mod = old_func.get_direct_user_nodes(lambda x: isinstance(x, Module))[0]
+            func.set_current_user_node(mod)
+
+        # Go back to the original Scope
+        self._scope = old_scope
+        self._current_function = old_current_function
+        # Retreive the function from the scope. For an imported function old_func remains unchanged.
+        # Otherwise this retrieves the semantic version
+        old_func = self.scope.find(old_func.name, 'functions')
+        # Remove the old_func from the imports dict and Assign the new annotated one
+        if old_func.is_imported:
```

Comment:
I still don't understand why we need to do `old_func = self.scope.find(old_func.name, 'functions')` instead of just using `old_func` directly. In what circumstances will the `is_imported` property change in this function?

Reply:
There is no need really to that line, so we can remove it

Reply:
[38ac876](https://github.com/pyccel/pyccel/pull/1728/commits/38ac876ba7507cc9e698e3e314427991fde8b87c)